edx-olive/configuration-old
# [SE-1194] Campus Automate clean-up of obsolete EBS volumes

This PR cherry picks two commits by Agrendalath which will run the log syncing as a systemd service. This ensures that except in cases of kernel panic or hard shutdown (pulling the plug) or network failures, tracking logs will always be rotated before an instance is terminated.

If this works, it will no longer be necessary to retain volumes after shutdown or run a separate job for log rotation.

**Testing instructions**

1. Go to the [stage asgard cluster](http://admin.edx-flatu.org/eu-west-1/cluster/show/stage-olivex-edxapp) using the instructions in the wiki for how to connect.

1. Create a cluster with the AMI of this branch `ami-01ae7f6743cd4ce76`. No need to put it in service.

1. Check that you can ssh to the instance created by the new cluster. Note the instance ID in the hostname and the IP -- you'll need them later.

1. Check that you can use s3cmd: `s3cmd  --access_key KEY_GOES_HERE --secret_key KEY_GOES_HERE ls s3://stage-olivex-tracking-logs/logs/tracking/stage-olivex-edxapp-sg/`

1. Check that you can log into AWS Campus -- you can do this from the awscli if you prefer, but you need access.

1. Wait until you are not near an hour boundary. We want to be sure that you are getting termination logs, not hourly rotation logs.

1. As root on the new instance, add some logs `echo TEST | sudo tee --append /edx/var/log/tracking/tracking.log`, `echo TEST2 | sudo tee --append /edx/var/log/tracking/tracking.log`.

1. Go to [Jenkins](https://admin.edx-flatu.org:8080/), wait until the `retire-instances-in-terminating-wait-prod-olivex` and `terminate-instances-that-have-been-verified-for-retirement-prod-olivex` jobs run and then disable them. Don't leave them off for more than about 10 minutes.

1. In Asgard, delete the cluster. In AWS, find and terminate the instance.

1. Grab the logs `s3cmd  --access_key KEY_GOES_HERE --secret_key KEY_GOES_HERE ls s3://stage-olivex-tracking-logs/logs/tracking/stage-olivex-edxapp-sg/<instance ID>-<IP>/`, `s3cmd  --access_key KEY_GOES_HERE --secret_key KEY_GOES_HERE get s3://stage-olivex-tracking-logs/logs/tracking/stage-olivex-edxapp-sg/<instance ID>-<IP>/$latest_file`. (For example: `s3://stage-olivex-tracking-logs/logs/tracking/stage-olivex-edxapp-sg/i-084fcb1e6e007172d-10.22.11.78/tracking.log-20190710-1562772552.gz`)

1. Grep the logs `zgrep TEST tracking.log-20190710-1562772552.gz`.

1. Turn the Jenkins jobs back on.

**Author's notes & concerns**

~I'm not sure how to test this or whether it is necessary to test it, since it is already merged into upstream master.~

Would it be sufficient to deploy this on stage by rebuilding the edxapp images and then turn off the log rotation jobs and verifying that logs are still being rotated on termination? Which jobs control the log rotation: are they in Jenkins?
